### ISSUE TITLE
Smarter selection of current working directory

### ISSUE THREAD

**Issue by nathancarter** [nteract/archived-desktop-app#4587]
**Is your feature request related to a problem? Please describe.**
In the nteract.app on Mac OS X (version 0.15.0), I'm always frustrated when I do File > New > Python 3 then File > Save As, but the Python runtime in the notebook is somehow in my home folder.  I never asked it to initialize its cwd to my home folder, I can't imagine a situation in which I would want that.  The only way to fix it is to close my new file and reopen it again from where I saved it.  Even Runtime > Restart and Run All Cells doesn't update the Python runtime to the correct folder.

**Describe the solution you'd like**
Two things:
 1. When saving a freshly created notebook for the first time, and no cells have ever been run, immediately change its Python runtime's working directory to the directory in which the notebook was just saved.
 2. When restarting the kernel, update its current working directory to the directory in which the notebook is saved.  (I thought #919 fixed this, but I guess not?)

**Describe alternatives you've considered**
Show cwd in status bar and let user click there to change it?

**Additional context**
Imagine also that someone would do a lot of work in the notebook, creating and saving files, not realizing that the output folder were their home folder.  Later, they re-open the notebook and it behaves differently, because now it has a different working directory than the last time they had it open.  An entirely different pain point from this same problem.  Also, #386 seems to address a similar issue, but it did not describe how it resolved it.

**Comment by captainsafia** [comment]
Thanks for opening this issue, @nathancarter!

Triage note: I think this is a dupe of an existing issue, other than #386 mentioned above, but cannot find it at the moment.

> When saving a freshly created notebook for the first time, and no cells have ever been run, immediately change its Python runtime's working directory to the directory in which the notebook was just saved.

When you launch the notebook that was just saved again, is the cwd set to the correct path?

> When restarting the kernel, update its current working directory to the directory in which the notebook is saved. (I thought #919 fixed this, but I guess not?)

The change in #919 doesn't affect the restart path so this is still a relevant issue.

I'm a little hesitant to take the approach in #1 as it requires us to implement state-dependent logic.

#2 should be resolved though and will likely need to be done in conjunction with #4575.

**Comment by nathancarter** [comment]
Hi, @captainsafia -- thanks for the quick reply!

> When you launch the notebook that was just saved again, is the cwd set to the correct path?

The cwd is correct only if you close the notebook entirely in nteract, then re-open it; in that case it is correct.  Just saving the file for the first time does not correctly set the cwd.

> I'm a little hesitant to take the approach in #1 as it requires us to implement state-dependent logic.

Yes, I see.  I probably shouldn't have suggested a specific method of solution, since I don't know your system or goals, sorry.  But here's what I would think nteract would like to avoid:
 1. Very common workflow that fails:
    * Create a notebook that will read some data file X and save it in the same folder as X.
    * Write some code and run it.
    * Can't find the file, you diagnose, and eventually find that, even though you never told it anything about your home directory, that's where it's running itself.  (Negative user surprise.)
 2. If the user who experiences that surprise want to keep using nteract:
    * They have to invent a workaround and use it every time in the future that they do this very common/natural workflow.
    * Right now that workaround actually involves closing and re-opening the notebook.
    * Even if you solve the "restart kernel means change cwd" problem, it's still a workaround that the user has to guess or figure out in some way and remember to employ every time, to avoid the unwanted behavior.

In short, "run this in my home folder" is a default setting that is wrong very often and hard to change.

**Comment by captainsafia** [comment]
When you open nteract, a blank notebook is created. This new notebook is opened in the users home directory. When you launch the Jupyter kernel, you must give it a path to launch the kernel under. It's convention to use the same path the notebook is under to launch the kernel, so that is what nteract does.

The problem happens when you save the notebook to a different location after it has been first created. At the moment, the notebook file is in a different directory but the kernel is still running under the directory that it was launched with in the first place.

I think the ideal solution would involve adding support for a "New notebook here" context menu option on browsers so that users can start off by creating the notebook in the location that the eventually wanted to reside in instead of having to go through the "Save As" escape hatch.

 So the default setting isn't "run this in my home folder" it's "run this in the same directory as this current notebook". For new notebooks, this ends up being the home folder.

**Comment by nathancarter** [comment]
That all makes sense, but speaks from the point of view of implementation challenges rather than user needs and expectations.  But perhaps I just don't know nteract's goals well enough:  I had been assuming that the goal is to wrap Jupyter in a standard app-like interface that behaves the way people expect apps to behave.  But perhaps the goal is actually an app that behaves the way that users who already know Jupyter expect Jupyter to behave?

If the goal is standard app behavior, then my concerns still make sense.  I cannot think of any software application that requires users to visit Windows Explorer/OS X Finder/etc. to create a new document, so proposing that as the correct solution for nteract users seems destined to create confusion and annoyance.  By contrast, nearly all applications use the new-doc-then-save-as workflow, and have done so for decades; won't all users expect that and be confused and annoyed if it doesn't work?

But I hear you regarding the particular implementation challenges here because the underlying engine is Jupyter, which behaves a certain way.  If we go with the current behavior, we have the annoyances I've described above.  If we automatically change working directory or restart the kernel upon Save As, we still get negative user surprise, just for a different reason.

But there is a standard way to resolve dilemmas like these:  When the user clicks Save As, you show a dialog box that explains the ambiguity and asks whether or not they want the kernel restarted in the new directory.  It can have the standard "Don't ask again; just do this every time" checkbox that such dialogs typically have.  Then each user gets to have the behavior they want, and the surprise is avoided because the dialog explicitly flags it.

On the other hand, if the goal is behavior similar to Jupyter, then my concerns make a little less sense.  But even so, I'm curious:  How do nteract staff create new files when using your own product?  I see only inconvenient options here, and it would surprise me to know that nteract devs themselves use one of these options and are happy with it.
 * Create the notebook outside nteract (e.g., use Jupyter or copy an old notebook) then open it in nteract afterwards
 * Put all your notebooks in your home folder (!), as the app seems to encourage (!!)
 * Start each notebook with an `import os` and then `os.chdir()` call (customized to the language in question) with absolute paths
 * The workflow I described earlier in this thread, closing and re-opening files

Aren't nteract devs themselves bothered by a list of new-file options that are all inconvenient?  Or am I just missing what the obvious correct and convenient workflow is?

**Comment by willingc** [comment]
I'm going to close this issue since for the current Jupyter kernel implementation that @captainsafia's response makes sense. Future implementations of the underlying Jupyter Server may improve the ability to offer additional file system flexibility.

Just reminding folks that everyone working on nteract is a volunteer. If someone from the community would like to champion working on the suggested changes, I would recommend opening a new issue as an RFC to discuss those changes and work effort to support.

**Comment by nathancarter** [comment]
Hi, @willingc -- sorry, I somehow didn't realize that nteract was not part of anyone's business, but a community project.  Not sure how I confused that, since it's pretty clear on the website!  My bad.

I'd be happy to open a new issue to discuss ways forward here, but I'm unfamiliar with opening an RFC...do I just start its title with "RFC:" and open a discussion, or is there a more formal procedure?  Thanks!

**Comment by captainsafia** [comment]
@nathancarter No formal procedure. An issue with the title "RFC" and a description of the problem and different proposed solutions will work fine. Thanks for offering to help!

**Comment by nathancarter** [comment]
Hmm...I just wrote up the issue, but before I submitted, I let GitHub compare it to existing issues, and it found #721, which is...well, exactly what I suggested on December 4.  And it seems it was solved back in 2018...but I'm not experiencing that behavior in my copy of nteract (v0.22.0).  Should I just file a bug against that feature instead?  Or was it intentionally removed?

**Comment by captainsafia** [comment]
That feel when someone sends you a link to a PR you wrote 3 years ago.

![image](https://media1.giphy.com/media/i30tbBjTNjPqg/giphy.gif?cid=790b7611ad7a434e97841d016ca82b2c087a96fb953e3c83&rid=giphy.gif)

It looks like that logic I added 3 years ago was indeed removed. I'm not sure why though. It's going to be hard to track the root cause since we converted our codebase from JavaScript to TypeScript a little over a year ago and GitHub doesn't provide the best tools for exploring changes across file renames.

My intuition is let's bring back this change under the assumption that it was unintentionally removed during the various refactors that have happened over the years.

Thanks for tracking this down, @nathancarter! It can be rather hard to keep track of years worth of issues and resolutions, :)

Since the change should involve copying over the code from the existing PR submitted three years ago, would you be interested in making this contribution? Happy to guide you through it on Slack.

Also, we can re-open this issue since it contains a reference to the previous issue and the PR that resolved it.

**Comment by nathancarter** [comment]
Sure, I'll try it.  You picked a good week...I'm on Spring Break.  How do I get into the Slack?  (If you need to send me an email, the address is in my profile.)

**Comment by captainsafia** [comment]
You can request an invite at slack.nteract.io and there is also a badge in the README of the repo.
